### PR TITLE
Add blog post on code coverage

### PR DESCRIPTION
**Pull request by d33bs**
This PR adds a blog post on code coverage, including some short examples and concepts of how it may be used through Python. 

Thanks for any feedback you may have on this!

### CODE REVIEW

File: _posts/2024-07-28-Uncovering-Code-Coverage.md [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+- **Function coverage** checks whether each function or method in the code has been called during the tests, ensuring that all parts of the code's interface are exercised.
+- **Branch coverage**, on the other hand, goes deeper by verifying whether every possible path through control structures (like if-else statements) has been followed, ensuring that all potential outcomes and decision points are tested. 
+
+## Benefits of Test Coverage
+
+> "If you are testing thoughtfully and well, I would expect a coverage percentage in the upper 80s or 90s. I would be suspicious of anything like 100% - it would smell of someone writing tests to make the coverage numbers happy, but not thinking about what they are doing." ([Martin Fowler, Test Coverage](https://martinfowler.com/bliki/TestCoverage.html))
+
+High test coverage ensures your code is reliable and less prone to bugs. 
+It helps you identify untested parts of your codebase, facilitating better maintenance and encouraging a culture of quality.
+This is balanced by understanding that 100% coverage is often very challenging to achieve and could be harmful in practice.
```
Comment (vincerubinetti) [inline _posts/2024-07-28-Uncovering-Code-Coverage.md:76]: Maybe the keyword "anti-pattern" could be thrown in somewhere in the post. I've seen a lot of people say 100% coverage is an anti-pattern that just wastes time, but not sure how much of a CS consensus that is.

File: _posts/2024-07-28-Uncovering-Code-Coverage.md [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+else:
+    return False
+```
+
+Line coverage, statement coverage, function coverage, and branch coverage common metrics used to measure how thoroughly code is tested, but they each focus on different aspects.
+
+- **Line coverage** tracks the execution of individual lines of code, indicating whether each line has been run during testing.
+- **Statement coverage** is similar but considers whether each statement (a single unit of execution) within the code has been executed, regardless of the number of lines.
+- **Function coverage** checks whether each function or method in the code has been called during the tests, ensuring that all parts of the code's interface are exercised.
+- **Branch coverage**, on the other hand, goes deeper by verifying whether every possible path through control structures (like if-else statements) has been followed, ensuring that all potential outcomes and decision points are tested.
```
Comment (vincerubinetti) [inline _posts/2024-07-28-Uncovering-Code-Coverage.md:68]: Might be overthinking this, but "on the other hand, goes deeper by" suggested to me that this was the  most thorough of all the options. If I understand correctly, line and statement coverage would always ensure all branches are covered.

Maybe reword to "Branch coverage goes deeper as compared to function coverage"... or just skip the preamble since the other bullets don't have it and just say "Branch coverage verifies whether".

File: _posts/2024-07-28-Uncovering-Code-Coverage.md [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+
+By doing so, developers can immediately see which parts of the code are covered by tests and which are not, enabling them to identify potential gaps in testing.
+This transparency helps maintain high code quality, ensures that new changes are well-tested, and encourages best practices in test-driven development.
+
+## Final Thoughts
+
+Code coverage is a vital part of software development that helps maintain code quality and reliability. By integrating coverage tools and practices into your workflow, you can ensure your code is thoroughly tested and ready for any changes. Start using Code coverage tools in your projects today and experience the benefits firsthand!
+
+## Additional Resources
+For further reading and resources, check out the documentation for .... There are also plenty of tutorials and best practices guides available online to help you get the most out of your Code coverage efforts.
```
Comment (vincerubinetti) [inline _posts/2024-07-28-Uncovering-Code-Coverage.md:224]: `....` needs to be filled in

**Review by vincerubinetti — APPROVED** [review]
Looks good, I especially appreciated the introduction sections, as code coverage is something I'm not experienced with.

File: _posts/2024-07-28-Uncovering-Code-Coverage.md [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+> "If you are testing thoughtfully and well, I would expect a coverage percentage in the upper 80s or 90s. I would be suspicious of anything like 100% - it would smell of someone writing tests to make the coverage numbers happy, but not thinking about what they are doing." ([Martin Fowler, Test Coverage](https://martinfowler.com/bliki/TestCoverage.html))
+
+High test coverage ensures your code is reliable and less prone to bugs. 
+It helps you identify untested parts of your codebase, facilitating better maintenance and encouraging a culture of quality.
+This is balanced by understanding that 100% coverage is often very challenging to achieve and could be harmful in practice.
+
+There are often times where you need to make changes to your code which can cause unforeseen issues to arise (for example, when there's coupling to other functionality).
+With good test coverage, you can more confidently refactor code during these occasions, knowing that any issues will be quickly caught by your tests.
+In short, it's a vital tool for any developer, providing you further insights about software which can otherwise go unseen.
+
```
Comment (the-mayer) [inline _posts/2024-07-28-Uncovering-Code-Coverage.md:81]: Just to really drive the point home, maybe consider a bullet about how implementing tests early in the development process is often easier than later. 

"Implementing code coverage at the beginning of the development process promotes writing cleaner and more reliable code."

**Review by the-mayer — APPROVED** [review]
Lots of great, digestible content! Outside the scope of this post, but designing the unit tests that contribute to code coverage is it's own can of worms... future blog post?

File: _posts/2024-07-28-Uncovering-Code-Coverage.md (hunk 1 above)
Comment (d33bs) [reply _posts/2024-07-28-Uncovering-Code-Coverage.md:76]: Thanks, I think that's appropriate, I'll work this in!

File: _posts/2024-07-28-Uncovering-Code-Coverage.md (hunk 2 above)
Comment (d33bs) [reply _posts/2024-07-28-Uncovering-Code-Coverage.md:68]: Thanks, I'll reword this for clarity. Generally I think each method goes deep or shallow in certain areas. Branch coverage does go a bit deeper with concerns to branches than statement-level coverage; if-else blocks are generally treated as one compound statement, meaning that when checking statement coverage any branch triggered could show the full if-else block as covered (despite there being branches which perhaps are untested).

File: _posts/2024-07-28-Uncovering-Code-Coverage.md (hunk 3 above)
Comment (d33bs) [reply _posts/2024-07-28-Uncovering-Code-Coverage.md:224]: Oops, thanks! Will fix this.

File: _posts/2024-07-28-Uncovering-Code-Coverage.md (hunk 4 above)
Comment (d33bs) [reply _posts/2024-07-28-Uncovering-Code-Coverage.md:81]: Thanks, love it, will add this in!

**Comment by d33bs** [comment]
Thanks a bunch @vincerubinetti and @the-mayer ! I've addressed your comments with further updates and will now merge this in.
